dex/pair/src/safe_price_view.rs

```diff
@@ -0,0 +1,422 @@
+        );
+        let start_round = current_round - round_offset;
+        self.get_safe_price(start_round, current_round, input_payment)
+    }
+
+    #[label("safe-price-view")]
+    #[view(getSafePriceByTimestampOffset)]
+    fn get_safe_price_by_timestamp_offset(
+        &self,
+        timestamp_offset: u64,
+        input_payment: EsdtTokenPayment<Self::Api>,
+    ) -> EsdtTokenPayment<Self::Api> {
+        let current_round = self.blockchain().get_block_round();
+        let round_offset = timestamp_offset / SECONDS_PER_ROUND;
+        require!(
+            round_offset > 0 && round_offset < current_round,
+            ERROR_PARAMETERS
+        );
+        let start_round = current_round - round_offset;
+        self.get_safe_price(start_round, current_round, input_payment)
+    }
+
+    #[label("safe-price-view")]
+    #[view(getSafePrice)]
+    fn get_safe_price(
+        &self,
+        start_round: Round,
+        end_round: Round,
+        input_payment: EsdtTokenPayment<Self::Api>,
+    ) -> EsdtTokenPayment<Self::Api> {
+        require!(end_round > start_round, ERROR_PARAMETERS);
+
+        let safe_price_current_index = self.safe_price_current_index().get();
+        let price_observations = self.price_observations();
+
+        let oldest_price_observation =
+            self.get_oldest_price_observation(safe_price_current_index, &price_observations);
+        require!(
+            oldest_price_observation.recording_round <= start_round,
+            ERROR_SAFE_PRICE_OBSERVATION_DOES_NOT_EXIST
+        );
+
+        let first_price_observation =
+            self.get_price_observation(safe_price_current_index, &price_observations, start_round);
+        let last_price_observation =
+            self.get_price_observation(safe_price_current_index, &price_observations, end_round);
+
+        self.compute_weighted_price(
+            input_payment,
+            &first_price_observation,
+            &last_price_observation,
+        )
+    }
+
+    #[label("safe-price-view")]
+    #[view(getPriceObservation)]
+    fn get_price_observation_view(&self, search_round: Round) -> PriceObservation<Self::Api> {
+        let safe_price_current_index = self.safe_price_current_index().get();
+        let price_observations = self.price_observations();
+
+        self.get_price_observation(safe_price_current_index, &price_observations, search_round)
+    }
+
+    fn compute_weighted_price(
+        &self,
+        input_payment: EsdtTokenPayment<Self::Api>,
+        first_price_observation: &PriceObservation<Self::Api>,
+        last_price_observation: &PriceObservation<Self::Api>,
+    ) -> EsdtTokenPayment<Self::Api> {
+        let first_token_id = self.first_token_id().get();
+        let second_token_id = self.second_token_id().get();
+
+        require!(
+            input_payment.token_identifier == first_token_id
+                || input_payment.token_identifier == second_token_id,
+            ERROR_BAD_INPUT_TOKEN
+        );
```

Comment:
To avoid comparing token ID twice, I'd say do this at the bottom. i.e.
```
if input_payment.token_identifier == first_token_id { ... }
else if input_payment.token_identifier == second_token_id { ... }
else { sc_panic!(...) }
```

Reply:
Updated.

Reply:
Remove the `require!`